### PR TITLE
Add cart page and mock data

### PR DESCRIPTION
**Pull request by Alex7Li**
A user needs to be able to see their cart when they go to the cart page, this pull request creates a barebones cart page for them. The cart uses mock data which is shared with the whole application by using the react redux library. This commit also populates the mock data and creates helper accessor functions which should be helpful for later development work. Probably this should have been two different pull requests, but I realized that a bit late.

Here is the page after adding in the bootstrap css link (will be done in #5)
Screenshot 2020-06-10 at 11 01 44 AM

In summary, we create the cart page, and populate it with fake data.

### CODE REVIEW

File: demo/src/store/store.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ * along with the quantity in the user's cart and if it is in the user's
+ * cart. If the item is not in the cart, the quantity is 0, but the converse
+ * is not always true, users that modify how many items are in their cart
+ * will usually not want it to disappear immediately.
+ */
+const startingState = {
+  items: {
+    '1jef2': {name: 'Airplane Ears', image: airplaneEars,
+      cost: 55, quantity: 1, inCart: true,
+      description: 'He wants to fly away with your heart'},
```
Comment (zikas) [inline demo/src/store/store.js:62]: Please use trailing commas everywhere
https://google.github.io/styleguide/jsguide.html#features-arrays-trailing-comma

File: demo/src/store/store.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+/**
+ * Functions to update the value saved in the store.
+ * @param {!Object} state The state to take in initially.
+ * @param {{type:string, id:string, amount:?number}} action a description of
+ * the action to take.
+ * @return {!Object} the new State
+ */
+function reducer(state = [], action) {
+  // deep copy state so as not to mutate anything
+  const newState = JSON.parse(JSON.stringify(state));
```
Comment (zikas) [inline demo/src/store/store.js:22]: things like this usually belong to utils where they could be properly tested and well documented. (deep copy using JSON.parse/stringify has its own edge-cases)

File: demo/src/store/store.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import makinBiscuits from '../images/makin_biscuits.jpg';
+import pawsUp from '../images/paws_up.jpg';
+import petMyBelly from '../images/pet_my_belly.jpg';
+import phoneBackground from '../images/phone_background.jpg';
+import snoozing from '../images/snoozing.jpg';
+import spoiled from '../images/spoiled.jpg';
+import whatYouLookinAt from '../images/what_you_lookin_at.jpg';
+
+/**
+ * Functions to update the value saved in the store.
```
Comment (zikas) [inline demo/src/store/store.js:14]: Per 
https://google.github.io/styleguide/jsguide.html#jsdoc-method-and-function-comments

Method descriptions begin with a verb phrase that describes what the method does. This phrase is not an imperative sentence, but instead is written in the third person, as if there is an implied This method ... before it.

File: demo/src/store/StoreHelpers.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  return {
+    type: 'DECREMENT',
+    id: id,
+  };
+}
+
+/**
+ * Helper function to set the number of a specific item in the cart. Does
+ * not change the inCart flag.
+ * @param {string} id: the id of the item to update
```
Comment (zikas) [inline demo/src/store/StoreHelpers.js:34]:
```suggestion
 * @param {string} id the id of the item to update
 * @param {number} amount the amount of items the user wants to buy
```

File: demo/src/store/StoreHelpers.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ */
+export function removeOneFromCart(id) {
+  return {
+    type: 'DECREMENT',
+    id: id,
+  };
+}
+
+/**
+ * Helper function to set the number of a specific item in the cart. Does
```
Comment (zikas) [inline demo/src/store/StoreHelpers.js:32]: Per 
https://google.github.io/styleguide/jsguide.html#jsdoc-method-and-function-comments

Method descriptions begin with a verb phrase that describes what the method does. This phrase is not an imperative sentence, but instead is written in the third person, as if there is an implied This method ... before it.

File: demo/src/store/StoreHelpers.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  return {
+    type: 'INCREMENT',
+    id: id,
+  };
+}
+
+/**
+ * Helper function to remove an item from the cart. If the quantity is
+ * set to zero, the inCart flag is automatically set to false.
+ * @param {string} id: the id of the item to update
```
Comment (zikas) [inline demo/src/store/StoreHelpers.js:21]:
```suggestion
 * @param {string} id the id of the item to update
```

File: demo/src/store/StoreHelpers.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ * @return {{type:string, id:string}} object describing the operation
+ */
+export function addOneToCart(id) {
+  return {
+    type: 'INCREMENT',
+    id: id,
+  };
+}
+
+/**
```
Comment (zikas) [inline demo/src/store/StoreHelpers.js:18]: Per 
https://google.github.io/styleguide/jsguide.html#jsdoc-method-and-function-comments

Method descriptions begin with a verb phrase that describes what the method does. This phrase is not an imperative sentence, but instead is written in the third person, as if there is an implied This method ... before it.

File: demo/src/store/StoreHelpers.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+ * Helper functions to modify the store
+ */
+
+/**
+ * Helper function to add an item from the cart and set the inCart
+ * flag to true.
+ * @param {string} id: the id of the item to update
```
Comment (zikas) [inline demo/src/store/StoreHelpers.js:8]:
```suggestion
 * @param {string} id the id of the item to update
```

File: demo/src/store/StoreHelpers.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
+/**
+ * Helper functions to modify the store
+ */
+
+/**
+ * Helper function to add an item from the cart and set the inCart
```
Comment (zikas) [inline demo/src/store/StoreHelpers.js:6]: Per 
https://google.github.io/styleguide/jsguide.html#jsdoc-method-and-function-comments

Method descriptions begin with a verb phrase that describes what the method does. This phrase is not an imperative sentence, but instead is written in the third person, as if there is an implied This method ... before it.

File: demo/src/store/StoreHelpers.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
+/**
```
Comment (zikas) [inline demo/src/store/StoreHelpers.js:1]: what this comment belongs to? The file itself? then please use
`@fileoverview` jsdoc
https://google.github.io/styleguide/jsguide.html#jsdoc-top-file-level-comments

File: demo/src/components/Cart/Cart.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
+import React from 'react';
+import {connect} from 'react-redux';
+import {Container, Col, Row, Form} from 'react-bootstrap';
+import './Cart.css';
+import {setQuantity} from '../../store/StoreHelpers';
```
Comment (zikas) [inline demo/src/components/Cart/Cart.js:5]: Per the styleguide:
The .js file extension is not optional in import paths and must always be included.

https://google.github.io/styleguide/jsguide.html#es-module-imports

File: demo/src/components/Cart/Cart.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+import Image from 'react-bootstrap/Image';
+import PropTypes from 'prop-types';
+
+/**
+ * Creates a component describing a shopping Cart.
+ * @param {!Object} state The site state
+ * @param {function} setQuantity function to modify quantity of items
```
Comment (zikas) [inline demo/src/components/Cart/Cart.js:12]: Are you using closure compiler? I believe valid function type definition must declare parameters, for example
@param {function(number): number}

File: demo/src/components/Cart/Cart.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+ *      in the global state
+ * @return {!JSX} The component.
+ */
```
Comment (zikas) [inline demo/src/components/Cart/Cart.js:15]: I guess it's just me not being very familiar with react but why do you need the name here instead of 
`const CartBase = function ({items, setQuantity}) {`

File: demo/src/components/Cart/Cart.js [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+const CartBase = function Cart({items, setQuantity}) {
+  const itemsRender = [];
```
Comment (zikas) [inline demo/src/components/Cart/Cart.js:17]: const /** !Array<!JSX> */ itemsRender = [];

https://google.github.io/styleguide/jsguide.html#jsdoc-template-parameter-types
Always specify template parameters. This way compiler can do a better job and it makes it easier for readers to understand what code does.

File: demo/src/components/Cart/Cart.js [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+  /*
+   * Create the content of the cart display, with one row per item.
```
Comment (zikas) [inline demo/src/components/Cart/Cart.js:20]: Please use single line comment // for one-line comments

File: demo/src/components/Cart/Cart.js [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    if (item.inCart) {
+      itemsRender.push(<Row key={itemID}>
+        <Col><Image className='image-holder' src={item.image}/></Col>
+        <Col><h3>{item.name}</h3><p>{item.description}</p>
+          <Form>
+            <Form.Group>
+              <Form.Label>Quantity</Form.Label>
+              <Form.Control type="number" value={item.quantity}
+                onChange={(event) => {
+                  setQuantity(itemID, +event.target.value);
```
Comment (zikas) [inline demo/src/components/Cart/Cart.js:32]: It's preferred (although not a part of the styleguide yet to use explicit conversion to number
`Number(event.target.value)`
not a requirement though, so I'll leave it up to you

File: demo/src/components/Cart/Cart.js [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+                }}/>
+            </Form.Group>
+          </Form>
+        </Col>
+        <Col>{item.cost.toFixed(2)}$</Col>
```
Comment (zikas) [inline demo/src/components/Cart/Cart.js:37]: I'm not very familiar with react, what this dollar sign means in the end?

File: demo/src/components/Cart/Cart.js [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  return (
+    <Container>
+      <Row key='cart-header' className="header-row">
+        <Col/>
+        <Col/>
+        <Col>Unit Cost</Col>
+      </Row>
+      {itemsRender }
```
Comment (zikas) [inline demo/src/components/Cart/Cart.js:49]: extra space

File: demo/src/components/Cart/Cart.js [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+  );
+};
+
+
+CartBase.propTypes = {
+  items: PropTypes.object,
+  setQuantity: PropTypes.func,
+};
+
+// pass in all of the state as props to cart
```
Comment (zikas) [inline demo/src/components/Cart/Cart.js:65]: It's usually recommended to keep format consistent, 
```suggestion
// Pass in all of the state as props to cart.
```

File: demo/src/components/Cart/Cart.js (hunk 11 above)
Comment (Alex7Li) [reply demo/src/components/Cart/Cart.js]: Fixed

File: demo/src/components/Cart/Cart.js (hunk 12 above)
Comment (Alex7Li) [reply demo/src/components/Cart/Cart.js]: I'm not using closure compiler right now, just compiling with react, I'll look into using it

File: demo/src/components/Cart/Cart.js (hunk 13 above)
Comment (Alex7Li) [reply demo/src/components/Cart/Cart.js:16]: That was a mistake, removed it

File: demo/src/components/Cart/Cart.js (hunk 14 above)
Comment (Alex7Li) [reply demo/src/components/Cart/Cart.js]: Fixed

File: demo/src/components/Cart/Cart.js (hunk 15 above)
Comment (Alex7Li) [reply demo/src/components/Cart/Cart.js]: Fixed

File: demo/src/components/Cart/Cart.js (hunk 16 above)
Comment (Alex7Li) [reply demo/src/components/Cart/Cart.js]: Switched it to Number

File: demo/src/components/Cart/Cart.js (hunk 17 above)
Comment (Alex7Li) [reply demo/src/components/Cart/Cart.js:36]: It's just plaintext, so the display will be something like 92.00$ instead of 92.00

File: demo/src/components/Cart/Cart.js (hunk 18 above)
Comment (Alex7Li) [reply demo/src/components/Cart/Cart.js]: Removed the space

File: demo/src/components/Cart/Cart.js (hunk 19 above)
Comment (Alex7Li) [reply demo/src/components/Cart/Cart.js]: Consistent-ified

File: demo/src/store/StoreHelpers.js (hunk 10 above)
Comment (Alex7Li) [reply demo/src/store/StoreHelpers.js:1]: Thanks, I added the fileoverview tag

File: demo/src/store/StoreHelpers.js (hunk 9 above)
Comment (Alex7Li) [reply demo/src/store/StoreHelpers.js]: Wow, I never would have caught that violation, fixed it.

File: demo/src/store/StoreHelpers.js (hunk 4 above)
Comment (Alex7Li) [reply demo/src/store/StoreHelpers.js]: Removed

File: demo/src/store/store.js (hunk 3 above)
Comment (Alex7Li) [reply demo/src/store/store.js]: Reworded

File: demo/src/store/store.js (hunk 2 above)
Comment (Alex7Li) [reply demo/src/store/store.js]: I moved this to a different file, I'm still using JSON.parse/stringify since it works for now

File: demo/src/store/store.js (hunk 1 above)
Comment (Alex7Li) [reply demo/src/store/store.js:64]: I might be missing something, but I think that this object is following the style guide, there is a comma after every newline

File: demo/src/store/StoreHelpers.js (hunk 6 above)
Comment (Alex7Li) [reply demo/src/store/StoreHelpers.js]: Removed

File: demo/src/store/StoreHelpers.js (hunk 7 above)
Comment (Alex7Li) [reply demo/src/store/StoreHelpers.js:17]: Reworded

File: demo/src/store/StoreHelpers.js (hunk 8 above)
Comment (Alex7Li) [reply demo/src/store/StoreHelpers.js]: Removed the colon

File: demo/src/utils.js [hunk 20]
```diff
@@ -0,0 +1,13 @@
+/**
+ * @fileoverview Utility functions for the demo app
```
[the remark this answers is not in the document]
Comment (Alex7Li) [reply demo/src/utils.js]: I'm not sure if you mean a period of another line with a *, I added both

File: demo/src/utils.js [hunk 21]
```diff
@@ -0,0 +1,13 @@
...
+ */
+
+/**
+ * Create a deep copy of the passed in data, which is expected
+ * to be shaped like the sample data in store.js.
+ * @param {!Object} data The data to copy.
+ * @return {!Object} A deep copy of the input.
+ */
+export function deepCopyStore(data) {
```
[the remark this answers is not in the document]
Comment (Alex7Li) [reply demo/src/utils.js]: Changed

File: demo/src/components/Cart/Cart.js [hunk 22]
```diff
@@ -1,24 +1,23 @@
...
-import Image from 'react-bootstrap/Image';
+import {setQuantity} from '../../store/StoreHelpers.js';
 import PropTypes from 'prop-types';
 
 /**
  * Creates a component describing a shopping Cart.
- * @param {!Object} state The site state
- * @param {function} setQuantity function to modify quantity of items
- *      in the global state
+ * @param {!Object.<string,
```
[the remark this answers is not in the document]
Comment (Alex7Li) [reply demo/src/components/Cart/Cart.js]: Fixed

File: demo/src/store/StoreHelpers.js [hunk 23]
```diff
@@ -29,12 +28,12 @@ export function removeOneFromCart(id) {
 }
 
 /**
- * Helper function to set the number of a specific item in the cart. Does
+ * Set the number of a specified item in the cart to a specified value. Does
```
[the remark this answers is not in the document]
Comment (Alex7Li) [reply demo/src/store/StoreHelpers.js]: The grammar has been corrected, I left the block syntax as is as it feels too spread out otherwise to me. I put the dollar sign on the other side.

File: demo/src/App.css [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+  height: 24px;
+  fill: currentColor;
+}
+
+.cat-logo {
+  width: 36px;
+  height: 36px;
+  fill: currentColor;
+  margin-right: 1rem;
+}
```
[the remark this answers is not in the document]
Comment (Alex7Li) [reply demo/src/App.css]: Done.
